In [1]:
# Colab에서 한글 폰트 설치
!apt-get -qq update
!apt-get -qq install -y fonts-nanum

import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.font_manager as fm
import shutil
import os

# matplotlib font cache 삭제
shutil.rmtree(mpl.get_cachedir(), ignore_errors=True)

# NanumGothic 직접 등록
font_path = "/usr/share/fonts/truetype/nanum/NanumGothic.ttf"
fm.fontManager.addfont(font_path)

mpl.rcParams['font.family'] = 'NanumGothic'
mpl.rcParams['axes.unicode_minus'] = False

print("NanumGothic font setting complete.")

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Selecting previously unselected package fonts-nanum.
(Reading database ... 122402 files and directories currently installed.)
Preparing to unpack .../fonts-nanum_20200506-1_all.deb ...
Unpacking fonts-nanum (20200506-1) ...
Setting up fonts-nanum (20200506-1) ...
Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...
NanumGothic font setting complete.


In [2]:
# =========================================================
# 0. 기본 설정
# =========================================================

import os
import gc
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

# import
train = pd.read_csv('/content/drive/MyDrive/2026_기상청_축산/hanwoo_train.csv')

In [3]:
# 1. 기본 크기 확인
print("shape:", train.shape)

# 2. 컬럼 확인
print(train.columns.tolist())

# 3. 데이터 타입 확인
display(train.info())

# 4. 앞부분 확인
display(train.head())

# 5. -99 계열 결측 표준화
train = train.replace([-99, -99.0, "-99", "-99.0"], np.nan)

# 6. 결측 개수 / 비율 확인
missing_summary = (
    pd.DataFrame({
        "missing_count": train.isna().sum(),
        "missing_rate": train.isna().mean() * 100,
        "dtype": train.dtypes
    })
    .sort_values("missing_rate", ascending=False)
)

display(missing_summary)

# 7. 타깃 분포 확인
target_col = "LAST_GRADE" if "LAST_GRADE" in train.columns else "Last_grade"

display(train[target_col].value_counts(dropna=False))
display(train[target_col].value_counts(normalize=True, dropna=False) * 100)

shape: (2408699, 23)
['sido', 'sigungu', 'eupmyeondong', 'stn', 'ABATT_DATE', 'JUDGE_DATE', 'JUDGE_SEX', 'WEIGHT', 'BACKFAT', 'REA', 'WINDEX', 'WGRADE', 'INSFAT', 'YUKSAK', 'FATSAK', 'TISSUE', 'GROWTH', 'COST_AMT', 'AGE', 'BIRTH_YMD', 'CATTLE_NO', 'FARM_UNIQUE_NO', 'LAST_GRADE']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2408699 entries, 0 to 2408698
Data columns (total 23 columns):
 #   Column          Dtype  
---  ------          -----  
 0   sido            object 
 1   sigungu         object 
 2   eupmyeondong    object 
 3   stn             int64  
 4   ABATT_DATE      object 
 5   JUDGE_DATE      object 
 6   JUDGE_SEX       object 
 7   WEIGHT          int64  
 8   BACKFAT         float64
 9   REA             float64
 10  WINDEX          float64
 11  WGRADE          object 
 12  INSFAT          float64
 13  YUKSAK          float64
 14  FATSAK          float64
 15  TISSUE          float64
 16  GROWTH          float64
 17  COST_AMT        float64
 18  AGE             int64 

None

,sido,sigungu,eupmyeondong,stn,ABATT_DATE,JUDGE_DATE,JUDGE_SEX,WEIGHT,BACKFAT,REA,...,YUKSAK,FATSAK,TISSUE,GROWTH,COST_AMT,AGE,BIRTH_YMD,CATTLE_NO,FARM_UNIQUE_NO,LAST_GRADE
0,강원특별자치도,정선군,북평면,563,2023-01-01,2023-01-02,암,218,3.0,44.0,...,6.0,7.0,5.0,9.0,-99.0,76,20160915,mI4i8G0BJ8kWD6gdm77RmTzyuIx6N2ZMaZ7wFXx3xb4=,hqh4Qmh5g+ymbYtqjAeKeA==,3B
1,강원특별자치도,평창군,진부면,560,2023-01-01,2023-01-02,거세,520,12.0,100.0,...,5.0,3.0,3.0,3.0,-99.0,32,20200504,tMgfi1p35taO9GH4XN4x0bfO8czy79B8V9NyczRuV+8=,N5qW6dZ91QWCMHg/1kPdlA==,1B
2,경상북도,영주시,이산면,837,2023-01-01,2023-01-02,거세,452,14.0,104.0,...,4.0,3.0,1.0,3.0,-99.0,29,20200803,mdsSD4/sB06U68LNT6P/GUwei2ISyIBZ+R04iLw2mX0=,5UbtU8TgV2yu3Y4/jxuNMQ==,1++B
3,강원특별자치도,평창군,진부면,560,2023-01-01,2023-01-02,거세,454,12.0,95.0,...,5.0,3.0,1.0,3.0,-99.0,32,20200510,LZWd4YatUD851LyC/LFXuxPW7KEXTwVWQONOwqrkLPI=,N5qW6dZ91QWCMHg/1kPdlA==,1++B
4,강원특별자치도,평창군,진부면,560,2023-01-01,2023-01-02,거세,462,10.0,78.0,...,5.0,3.0,3.0,3.0,-99.0,32,20200511,A4FF/dhr8vg7JeW/eqc3922pA69nEaLL/+uoAusZqyU=,N5qW6dZ91QWCMHg/1kPdlA==,1B


,missing_count,missing_rate,dtype
COST_AMT,885003,36.741951,float64
INSFAT,5331,0.221323,float64
TISSUE,5320,0.220866,float64
YUKSAK,5269,0.218749,float64
WINDEX,5252,0.218043,float64
REA,5188,0.215386,float64
WGRADE,4814,0.199859,object
FATSAK,4743,0.196911,float64
BACKFAT,4567,0.189604,float64
GROWTH,9,0.000374,float64


,count
LAST_GRADE,
1++B,319588
1+B,311222
1B,299290
1++A,207026
2B,195866
1A,168174
1+A,167246
1+C,130094
1++C,128997


,proportion
LAST_GRADE,
1++B,13.268075
1+B,12.920751
1B,12.425380
1++A,8.594930
2B,8.131610
1A,6.981943
1+A,6.943416
1+C,5.401007
1++C,5.355464


## 1. 결측치 확인

In [4]:
# 여러 변수의 결측 행이 같은 행인지 확인
missing_cols = [
    "INSFAT", "TISSUE", "YUKSAK", "WINDEX", "REA",
    "WGRADE", "FATSAK", "BACKFAT", "GROWTH"
]

missing_pattern = (
    train[missing_cols]
    .isna()
    .astype(int)
    .value_counts()
    .reset_index(name="n")
)

display(missing_pattern.head(30))

,INSFAT,TISSUE,YUKSAK,WINDEX,REA,WGRADE,FATSAK,BACKFAT,GROWTH,n
0,0,0,0,0,0,0,0,0,0,2403291
1,1,1,1,1,1,1,1,1,0,4148
2,1,1,1,1,1,1,0,0,0,461
3,1,1,1,1,1,0,1,1,0,322
4,1,1,1,0,0,0,1,0,0,124
5,1,1,1,1,1,1,1,0,0,95
6,0,0,0,1,0,0,0,0,0,48
7,1,1,1,1,1,1,0,1,0,33
8,1,1,0,1,1,1,0,0,0,20
9,1,0,0,1,1,1,0,1,0,18


In [5]:
# 결측 행이 등외와 얼마나 연관이 있나
for col in missing_cols:
    print("\n", "="*80)
    print(col)
    display(train.loc[train[col].isna(), "LAST_GRADE"].value_counts(dropna=False))


INSFAT


,count
LAST_GRADE,
등외,5331



TISSUE


,count
LAST_GRADE,
등외,5320



YUKSAK


,count
LAST_GRADE,
등외,5269



WINDEX


,count
LAST_GRADE,
등외,5252



REA


,count
LAST_GRADE,
등외,5188



WGRADE


,count
LAST_GRADE,
등외,4814



FATSAK


,count
LAST_GRADE,
등외,4743



BACKFAT


,count
LAST_GRADE,
등외,4567



GROWTH


,count
LAST_GRADE,
등외,9


In [6]:
outgrade = train["LAST_GRADE"].eq("등외")

outgrade_missing_summary = pd.DataFrame({
    "missing_count_in_outgrade": train.loc[outgrade, missing_cols].isna().sum(),
    "outgrade_missing_rate": train.loc[outgrade, missing_cols].isna().mean() * 100,
    "total_missing_count": train[missing_cols].isna().sum()
})

display(outgrade_missing_summary.sort_values("outgrade_missing_rate", ascending=False))

,missing_count_in_outgrade,outgrade_missing_rate,total_missing_count
INSFAT,5331,96.227437,5331
TISSUE,5320,96.028881,5320
YUKSAK,5269,95.108303,5269
WINDEX,5252,94.801444,5252
REA,5188,93.646209,5188
WGRADE,4814,86.895307,4814
FATSAK,4743,85.613718,4743
BACKFAT,4567,82.436823,4567
GROWTH,9,0.162455,9


## 2. 결측치 처리 - 예측용


1. COST_AMT는 제외
2. 결측 행 삭제는 하지 않음
3. 변수 제거도 하지 않음
4. 원본값_raw 생성
5. 결측 여부_missing 생성
6. 수치형/순서형 변수는 group median 대치
7. WGRADE는 group mode 대치
8. 순서형 변수는 대치 후 정수 복원
9. impute_method를 남겨 fallback이 얼마나 발생했는지 확인

In [7]:
import numpy as np
import pandas as pd

# ============================================================
# 0. 원본 보존
# ============================================================

train_raw = train.copy()
train_imp = train.copy()

# -99 계열 sentinel이 있으면 NaN으로 통일
sentinel_values = [-99, -99.0, "-99", "-99.0"]
train_imp = train_imp.replace(sentinel_values, np.nan)
train_raw = train_raw.replace(sentinel_values, np.nan)

# ============================================================
# 1. 결측 처리 대상 변수 정의
# ============================================================

# 수치형 변수
numeric_impute_cols = [
    "INSFAT", "TISSUE", "YUKSAK", "FATSAK",
    "WINDEX", "REA", "BACKFAT", "GROWTH"
]

# 범주형 변수
categorical_impute_cols = [
    "WGRADE"
]

# 실제 존재하는 컬럼만 사용
numeric_impute_cols = [c for c in numeric_impute_cols if c in train_imp.columns]
categorical_impute_cols = [c for c in categorical_impute_cols if c in train_imp.columns]

# ============================================================
# 2. group key 생성
# ============================================================

# 판정연도 생성
if "JUDGE_DATE" in train_imp.columns:
    train_imp["JUDGE_DATE"] = pd.to_datetime(train_imp["JUDGE_DATE"], errors="coerce")
    train_imp["JUDGE_YEAR"] = train_imp["JUDGE_DATE"].dt.year
elif "ABATT_DATE" in train_imp.columns:
    train_imp["ABATT_DATE"] = pd.to_datetime(train_imp["ABATT_DATE"], errors="coerce")
    train_imp["JUDGE_YEAR"] = train_imp["ABATT_DATE"].dt.year
else:
    raise ValueError("JUDGE_DATE 또는 ABATT_DATE가 필요합니다.")

# 연령대 생성
train_imp["age_bin"] = pd.cut(
    train_imp["AGE"],
    bins=[0, 24, 30, 36, 42, 60, 999],
    labels=["<=24", "25-30", "31-36", "37-42", "43-60", "60+"],
    right=True
)

# group key
group_levels = [
    ["JUDGE_SEX", "age_bin", "sido", "JUDGE_YEAR"],
    ["JUDGE_SEX", "age_bin", "JUDGE_YEAR"],
    ["JUDGE_SEX", "JUDGE_YEAR"],
    ["JUDGE_SEX"]
]

In [8]:
# ============================================================
# 3. 수치형 변수 group median 대치 함수
# ============================================================

def hierarchical_group_median_impute(df, col, group_levels):
    """
    원본 관측값만 기준으로 group median을 계산하고,
    결측값을 계층적으로 대치합니다.
    """
    raw_s = df[col].copy()
    filled = raw_s.copy()

    missing_flag_col = f"{col}_missing"
    method_col = f"{col}_impute_method"
    raw_col = f"{col}_raw"

    df[raw_col] = raw_s
    df[missing_flag_col] = raw_s.isna().astype(int)
    df[method_col] = np.where(raw_s.isna(), "missing_not_imputed", "observed")

    for i, group_cols in enumerate(group_levels, start=1):
        # 원본 관측값 기준 median
        group_median = raw_s.groupby(
            [df[g] for g in group_cols],
            observed=True
        ).transform("median")

        fill_mask = filled.isna() & group_median.notna()

        filled.loc[fill_mask] = group_median.loc[fill_mask]
        df.loc[fill_mask, method_col] = f"group_median_level_{i}"

    # 최종 fallback: 전체 median
    global_median = raw_s.median()
    fill_mask = filled.isna()

    filled.loc[fill_mask] = global_median
    df.loc[fill_mask, method_col] = "global_median"

    df[col] = filled

    return df


# ============================================================
# 4. 수치형 변수 대치 실행
# ============================================================

for col in numeric_impute_cols:
    train_imp = hierarchical_group_median_impute(
        train_imp,
        col,
        group_levels
    )

In [9]:
# ============================================================
# 5. 순서형 변수 정수 복원
# ============================================================

ordinal_cols = [
    "INSFAT", "TISSUE", "YUKSAK", "FATSAK"
]

ordinal_cols = [c for c in ordinal_cols if c in train_imp.columns]

for col in ordinal_cols:
    missing_flag_col = f"{col}_missing"

    if missing_flag_col in train_imp.columns:
        imp_mask = train_imp[missing_flag_col].eq(1)

        # 일반적인 반올림: x.5 이상은 올림
        train_imp.loc[imp_mask, col] = np.floor(train_imp.loc[imp_mask, col] + 0.5)

        # 원본 관측값 범위를 벗어나지 않게 clipping
        obs_min = train_raw[col].min()
        obs_max = train_raw[col].max()

        train_imp.loc[imp_mask, col] = train_imp.loc[imp_mask, col].clip(obs_min, obs_max)

In [10]:
# ============================================================
# 6. 범주형 변수 group mode 대치 함수
# ============================================================

def get_group_mode_series(df, raw_s, group_cols):
    """
    각 group별 최빈값을 계산해서 원래 df 순서에 맞는 Series로 반환합니다.
    대용량 데이터에서 transform(lambda mode)보다 안정적입니다.
    """
    tmp = df.loc[raw_s.notna(), group_cols].copy()
    tmp["_value"] = raw_s.loc[raw_s.notna()].values

    mode_table = (
        tmp
        .groupby(group_cols + ["_value"], observed=True)
        .size()
        .reset_index(name="_n")
        .sort_values(group_cols + ["_n"], ascending=[True] * len(group_cols) + [False])
        .drop_duplicates(group_cols)
        .drop(columns="_n")
    )

    mode_series = (
        df[group_cols]
        .merge(mode_table, on=group_cols, how="left")["_value"]
    )

    mode_series.index = df.index

    return mode_series


def hierarchical_group_mode_impute(df, col, group_levels):
    raw_s = df[col].copy()
    filled = raw_s.copy()

    missing_flag_col = f"{col}_missing"
    method_col = f"{col}_impute_method"
    raw_col = f"{col}_raw"

    df[raw_col] = raw_s
    df[missing_flag_col] = raw_s.isna().astype(int)
    df[method_col] = np.where(raw_s.isna(), "missing_not_imputed", "observed")

    for i, group_cols in enumerate(group_levels, start=1):
        group_mode = get_group_mode_series(df, raw_s, group_cols)

        fill_mask = filled.isna() & group_mode.notna()

        filled.loc[fill_mask] = group_mode.loc[fill_mask]
        df.loc[fill_mask, method_col] = f"group_mode_level_{i}"

    # 최종 fallback: 전체 mode
    global_mode = raw_s.mode(dropna=True).iloc[0]
    fill_mask = filled.isna()

    filled.loc[fill_mask] = global_mode
    df.loc[fill_mask, method_col] = "global_mode"

    df[col] = filled

    return df


# ============================================================
# 7. 범주형 변수 대치 실행
# ============================================================

for col in categorical_impute_cols:
    train_imp = hierarchical_group_mode_impute(
        train_imp,
        col,
        group_levels
    )

In [11]:
# ============================================================
# 8. 결측 처리 후 확인
# ============================================================

check_cols = numeric_impute_cols + categorical_impute_cols

print("대치 후 결측 개수")
display(train_imp[check_cols].isna().sum())

print("변수별 대치 method 분포")
for col in check_cols:
    method_col = f"{col}_impute_method"
    if method_col in train_imp.columns:
        print("\n", "=" * 80)
        print(col)
        display(train_imp[method_col].value_counts(dropna=False))

대치 후 결측 개수


,0
INSFAT,0
TISSUE,0
YUKSAK,0
FATSAK,0
WINDEX,0
REA,0
BACKFAT,0
GROWTH,0
WGRADE,0


변수별 대치 method 분포

INSFAT


,count
INSFAT_impute_method,
observed,2403368
group_median_level_1,5329
group_median_level_2,2



TISSUE


,count
TISSUE_impute_method,
observed,2403379
group_median_level_1,5318
group_median_level_2,2



YUKSAK


,count
YUKSAK_impute_method,
observed,2403430
group_median_level_1,5267
group_median_level_2,2



FATSAK


,count
FATSAK_impute_method,
observed,2403956
group_median_level_1,4741
group_median_level_2,2



WINDEX


,count
WINDEX_impute_method,
observed,2403447
group_median_level_1,5250
group_median_level_2,2



REA


,count
REA_impute_method,
observed,2403511
group_median_level_1,5186
group_median_level_2,2



BACKFAT


,count
BACKFAT_impute_method,
observed,2404132
group_median_level_1,4565
group_median_level_2,2



GROWTH


,count
GROWTH_impute_method,
observed,2408690
group_median_level_1,9



WGRADE


,count
WGRADE_impute_method,
observed,2403885
group_mode_level_1,4814


In [12]:
# ============================================================
# 9. 대치 전후 요약 통계 비교
# ============================================================

for col in numeric_impute_cols:
    print("\n", "=" * 80)
    print(col)

    compare = pd.DataFrame({
        "raw_observed": train_raw[col].dropna().describe(),
        "imputed_all": train_imp[col].describe()
    })

    display(compare)


INSFAT


,raw_observed,imputed_all
count,2.403368e+06,2.408699e+06
mean,5.423036e+00,5.419405e+00
std,2.199511e+00,2.199205e+00
min,1.000000e+00,1.000000e+00
25%,4.000000e+00,4.000000e+00
50%,6.000000e+00,6.000000e+00
75%,7.000000e+00,7.000000e+00
max,9.000000e+00,9.000000e+00



TISSUE


,raw_observed,imputed_all
count,2.403379e+06,2.408699e+06
mean,2.183306e+00,2.184999e+00
std,1.061440e+00,1.061309e+00
min,1.000000e+00,1.000000e+00
25%,1.000000e+00,1.000000e+00
50%,2.000000e+00,2.000000e+00
75%,3.000000e+00,3.000000e+00
max,5.000000e+00,5.000000e+00



YUKSAK


,raw_observed,imputed_all
count,2.403430e+06,2.408699e+06
mean,4.967826e+00,4.967931e+00
std,4.916512e-01,4.911537e-01
min,1.000000e+00,1.000000e+00
25%,5.000000e+00,5.000000e+00
50%,5.000000e+00,5.000000e+00
75%,5.000000e+00,5.000000e+00
max,8.000000e+00,8.000000e+00



FATSAK


,raw_observed,imputed_all
count,2.403956e+06,2.408699e+06
mean,3.067800e+00,3.067667e+00
std,3.704607e-01,3.701089e-01
min,1.000000e+00,1.000000e+00
25%,3.000000e+00,3.000000e+00
50%,3.000000e+00,3.000000e+00
75%,3.000000e+00,3.000000e+00
max,8.000000e+00,8.000000e+00



WINDEX


,raw_observed,imputed_all
count,2.403447e+06,2.408699e+06
mean,6.151405e+01,6.151547e+01
std,1.824431e+00,1.824517e+00
min,2.919000e+01,2.919000e+01
25%,6.040000e+01,6.041000e+01
50%,6.158000e+01,6.158000e+01
75%,6.263000e+01,6.263000e+01
max,7.732000e+01,7.732000e+01



REA


,raw_observed,imputed_all
count,2.403511e+06,2.408699e+06
mean,9.312451e+01,9.310506e+01
std,1.506140e+01,1.505421e+01
min,1.000000e+00,1.000000e+00
25%,8.400000e+01,8.400000e+01
50%,9.300000e+01,9.300000e+01
75%,1.030000e+02,1.020000e+02
max,1.960000e+02,1.960000e+02



BACKFAT


,raw_observed,imputed_all
count,2.404132e+06,2.408699e+06
mean,1.263125e+01,1.262731e+01
std,5.453570e+00,5.450186e+00
min,1.000000e+00,1.000000e+00
25%,9.000000e+00,9.000000e+00
50%,1.200000e+01,1.200000e+01
75%,1.500000e+01,1.500000e+01
max,9.800000e+01,9.800000e+01



GROWTH


,raw_observed,imputed_all
count,2.408690e+06,2.408699e+06
mean,4.176861e+00,4.176864e+00
std,2.145014e+00,2.145019e+00
min,1.000000e+00,1.000000e+00
25%,3.000000e+00,3.000000e+00
50%,3.000000e+00,3.000000e+00
75%,6.000000e+00,6.000000e+00
max,9.000000e+00,9.000000e+00


첨부한 전후 요약을 보면 raw_observed와 imputed_all의 평균, 표준편차, 사분위수, 최솟값/최댓값이 거의 유지됩니다. 예를 들어 INSFAT 평균은 5.423 → 5.419, BACKFAT 평균은 12.631 → 12.627, GROWTH는 거의 변화가 없습니다. REA도 평균은 93.124 → 93.105로 거의 같고, 75% 분위수만 103 → 102로 아주 작게 바뀐 정도입니다. 결측률이 낮고 group median 대치라서 전체 분포를 크게 훼손하지 않은 것으로 보면 됩니다.

## 3. COST_AMT 결측치 처리

In [13]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from lightgbm import LGBMRegressor

# ============================================================
# 0. 기본 설정
# ============================================================

df = train_imp.copy()

required_cols = ["COST_AMT", "WEIGHT"]
missing_required = [c for c in required_cols if c not in df.columns]

if missing_required:
    raise ValueError(f"필요한 컬럼이 없습니다: {missing_required}")

# 원본 보존
df["COST_AMT_raw"] = df["COST_AMT"]
df["COST_AMT_missing"] = df["COST_AMT"].isna().astype(int)

# 비정상 값 처리
df.loc[df["WEIGHT"] <= 0, "WEIGHT"] = np.nan
df.loc[df["COST_AMT"] <= 0, "COST_AMT"] = np.nan

# kg당 가격 생성
df["COST_PER_WEIGHT_raw"] = df["COST_AMT"] / df["WEIGHT"]

# inf 제거
df["COST_PER_WEIGHT_raw"] = df["COST_PER_WEIGHT_raw"].replace([np.inf, -np.inf], np.nan)

In [14]:
# ============================================================
# 1. feature 후보 설정
# ============================================================

candidate_features = [
    # 기본 개체 정보
    "AGE", "WEIGHT", "GROWTH",

    # 성별/지역/시점
    "JUDGE_SEX", "sido", "sigungu", "eupmyeondong", "stn", "JUDGE_YEAR",

    # 등급/판정 관련 변수
    "LAST_GRADE", "WGRADE",
    "BACKFAT", "REA", "INSFAT", "TISSUE", "YUKSAK", "FATSAK", "WINDEX",

    # 결측 flag
    "BACKFAT_missing", "REA_missing", "INSFAT_missing", "TISSUE_missing",
    "YUKSAK_missing", "FATSAK_missing", "WINDEX_missing", "WGRADE_missing",
    "GROWTH_missing"
]

feature_cols = [c for c in candidate_features if c in df.columns]

# target 관련 변수는 혹시 들어가 있으면 제거
for bad_col in ["COST_AMT", "COST_AMT_raw", "COST_AMT_imp",
                "COST_PER_WEIGHT_raw", "COST_PER_WEIGHT_imp"]:
    if bad_col in feature_cols:
        feature_cols.remove(bad_col)

print("사용 feature 수:", len(feature_cols))
print(feature_cols)

사용 feature 수: 27
['AGE', 'WEIGHT', 'GROWTH', 'JUDGE_SEX', 'sido', 'sigungu', 'eupmyeondong', 'stn', 'JUDGE_YEAR', 'LAST_GRADE', 'WGRADE', 'BACKFAT', 'REA', 'INSFAT', 'TISSUE', 'YUKSAK', 'FATSAK', 'WINDEX', 'BACKFAT_missing', 'REA_missing', 'INSFAT_missing', 'TISSUE_missing', 'YUKSAK_missing', 'FATSAK_missing', 'WINDEX_missing', 'WGRADE_missing', 'GROWTH_missing']


In [15]:
# ============================================================
# 2. 범주형 변수 category 변환
# ============================================================

cat_cols = [
    "JUDGE_SEX", "sido", "sigungu", "eupmyeondong",
    "LAST_GRADE", "WGRADE"
]

cat_cols = [c for c in cat_cols if c in feature_cols]

for col in cat_cols:
    df[col] = df[col].astype("category")

In [16]:
# ============================================================
# 3. 관측 행 기준 train/valid 분리
# ============================================================

obs_mask = df["COST_PER_WEIGHT_raw"].notna() & df["WEIGHT"].notna()

X = df.loc[obs_mask, feature_cols].copy()
y = np.log1p(df.loc[obs_mask, "COST_PER_WEIGHT_raw"])

X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# ============================================================
# 4. LightGBM 학습
# ============================================================

cost_model = LGBMRegressor(
    n_estimators=800,
    learning_rate=0.05,
    max_depth=-1,
    num_leaves=63,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    n_jobs=-1
)

cost_model.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="rmse",
    categorical_feature=cat_cols
)

# ============================================================
# 5. 검증 성능 확인
# ============================================================

valid_pred_log = cost_model.predict(X_valid)
valid_pred = np.expm1(valid_pred_log)
valid_true = np.expm1(y_valid)

mae = mean_absolute_error(valid_true, valid_pred)
rmse = np.sqrt(mean_squared_error(valid_true, valid_pred))
r2 = r2_score(valid_true, valid_pred)

print("COST_PER_WEIGHT validation")
print("MAE :", mae)
print("RMSE:", rmse)
print("R2  :", r2)

[LightGBM] [Warning] Categorical features with more bins than the configured maximum bin number found.
[LightGBM] [Warning] For categorical features, max_bin and max_bin_by_feature may be ignored with a large number of categories.
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.114070 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2406
[LightGBM] [Info] Number of data points in the train set: 1218956, number of used features: 26
[LightGBM] [Info] Start training from score 3.692771
COST_PER_WEIGHT validation
MAE : 2.88521463165097
RMSE: 3.791982701841444
R2  : 0.8207674156971396


In [17]:
# ============================================================
# 6. 전체 관측 행으로 최종 모델 재학습
# ============================================================

X_all = df.loc[obs_mask, feature_cols].copy()
y_all = np.log1p(df.loc[obs_mask, "COST_PER_WEIGHT_raw"])

final_cost_model = LGBMRegressor(
    n_estimators=800,
    learning_rate=0.05,
    max_depth=-1,
    num_leaves=63,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    n_jobs=-1
)

final_cost_model.fit(
    X_all,
    y_all,
    categorical_feature=cat_cols
)

# ============================================================
# 7. COST_AMT 결측 행 예측
# ============================================================

miss_mask = df["COST_AMT_raw"].isna() & df["WEIGHT"].notna()

X_miss = df.loc[miss_mask, feature_cols].copy()

pred_cost_per_weight = np.expm1(final_cost_model.predict(X_miss))

# 비정상 예측 방지
pred_cost_per_weight = np.maximum(pred_cost_per_weight, 0)

# 대치 컬럼 생성
df["COST_PER_WEIGHT_imp"] = df["COST_PER_WEIGHT_raw"]
df["COST_AMT_imp"] = df["COST_AMT_raw"]
df["COST_AMT_impute_method"] = np.where(
    df["COST_AMT_raw"].isna(),
    "missing_not_imputed",
    "observed"
)

df.loc[miss_mask, "COST_PER_WEIGHT_imp"] = pred_cost_per_weight
df.loc[miss_mask, "COST_AMT_imp"] = (
    df.loc[miss_mask, "COST_PER_WEIGHT_imp"] * df.loc[miss_mask, "WEIGHT"]
)

df.loc[miss_mask, "COST_AMT_impute_method"] = "lightgbm_cost_per_weight"

[LightGBM] [Warning] Categorical features with more bins than the configured maximum bin number found.
[LightGBM] [Warning] For categorical features, max_bin and max_bin_by_feature may be ignored with a large number of categories.
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.446088 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 2402
[LightGBM] [Info] Number of data points in the train set: 1523695, number of used features: 26
[LightGBM] [Info] Start training from score 3.692698


In [18]:
# ============================================================
# 8. fallback: COST_PER_WEIGHT group median
# WEIGHT가 없거나 비정상이라 LightGBM으로 복원이 안 된 행이 있을 수 있습니다. 이런 행은 group median으로 fallback 처리
# ============================================================

fallback_mask = df["COST_AMT_imp"].isna()

group_levels = [
    ["JUDGE_SEX", "age_bin", "sido", "JUDGE_YEAR"],
    ["JUDGE_SEX", "age_bin", "JUDGE_YEAR"],
    ["JUDGE_SEX", "JUDGE_YEAR"],
    ["JUDGE_SEX"]
]

for i, group_cols in enumerate(group_levels, start=1):
    group_cols = [c for c in group_cols if c in df.columns]

    if len(group_cols) == 0:
        continue

    group_median = df.groupby(group_cols, observed=True)["COST_PER_WEIGHT_raw"].transform("median")

    fill_mask = fallback_mask & group_median.notna() & df["WEIGHT"].notna()

    df.loc[fill_mask, "COST_PER_WEIGHT_imp"] = group_median.loc[fill_mask]
    df.loc[fill_mask, "COST_AMT_imp"] = (
        df.loc[fill_mask, "COST_PER_WEIGHT_imp"] * df.loc[fill_mask, "WEIGHT"]
    )
    df.loc[fill_mask, "COST_AMT_impute_method"] = f"group_median_cost_per_weight_level_{i}"

    fallback_mask = df["COST_AMT_imp"].isna()

# 최종 fallback: 전체 kg당 가격 median
global_cost_per_weight_median = df["COST_PER_WEIGHT_raw"].median()

fill_mask = df["COST_AMT_imp"].isna() & df["WEIGHT"].notna()

df.loc[fill_mask, "COST_PER_WEIGHT_imp"] = global_cost_per_weight_median
df.loc[fill_mask, "COST_AMT_imp"] = (
    df.loc[fill_mask, "COST_PER_WEIGHT_imp"] * df.loc[fill_mask, "WEIGHT"]
)
df.loc[fill_mask, "COST_AMT_impute_method"] = "global_median_cost_per_weight"

In [37]:
# ============================================================
# 9. 최종 점검
# ============================================================

print("COST_AMT 원본 결측 수:", df["COST_AMT_raw"].isna().sum())
print("COST_AMT_imp 결측 수:", df["COST_AMT_imp"].isna().sum())
print("COST_PER_WEIGHT_imp 결측 수:", df["COST_PER_WEIGHT_imp"].isna().sum())

display(df["COST_AMT_impute_method"].value_counts(dropna=False))

# 대치 전후 요약
cost_compare = pd.DataFrame({
    "COST_AMT_raw_observed": df["COST_AMT_raw"].dropna().describe(),
    "COST_AMT_imp_all": df["COST_AMT_imp"].describe(),
    "COST_PER_WEIGHT_raw_observed": df["COST_PER_WEIGHT_raw"].dropna().describe(),
    "COST_PER_WEIGHT_imp_all": df["COST_PER_WEIGHT_imp"].describe()
})

display(cost_compare)

COST_AMT 원본 결측 수: 885004
COST_AMT_imp 결측 수: 0
COST_PER_WEIGHT_imp 결측 수: 0


,count
COST_AMT_impute_method,
observed,1523695
lightgbm_cost_per_weight,885003
global_median_cost_per_weight_final_fallback,1


,COST_AMT_raw_observed,COST_AMT_imp_all,COST_PER_WEIGHT_raw_observed,COST_PER_WEIGHT_imp_all
count,1.523695e+06,2.408699e+06,1.523695e+06,2.408699e+06
mean,1.666182e+04,1.678204e+04,4.013795e+01,4.017777e+01
std,4.212804e+03,3.894477e+03,8.964789e+00,8.601832e+00
min,1.000000e+02,1.000000e+02,2.702703e+00,2.702703e+00
25%,1.390000e+04,1.434500e+04,3.406326e+01,3.433791e+01
50%,1.694900e+04,1.706573e+04,3.963581e+01,3.973923e+01
75%,1.940000e+04,1.930000e+04,4.562771e+01,4.551204e+01
max,1.500000e+05,1.500000e+05,2.419355e+02,2.419355e+02


In [21]:
train_imp = df.copy()

In [38]:
# 문제 1 : COST_PER_WEIGHT_imp 1개 결측값 처리
fill_mask = df["COST_PER_WEIGHT_imp"].isna()

df.loc[fill_mask, "COST_PER_WEIGHT_imp"] = df["COST_PER_WEIGHT_imp"].median()

df.loc[fill_mask, "COST_AMT_imp"] = (
    df.loc[fill_mask, "COST_PER_WEIGHT_imp"] * df.loc[fill_mask, "WEIGHT"]
)

df.loc[fill_mask, "COST_AMT_impute_method"] = "global_median_cost_per_weight_final_fallback"

In [39]:
# 문제 2 : COST_AMT 최솟값이 0임

print("COST_AMT_raw == 0:", (df["COST_AMT_raw"] == 0).sum())
print("COST_AMT_imp == 0:", (df["COST_AMT_imp"] == 0).sum())

display(
    df.loc[df["COST_AMT_raw"] == 0,
           ["COST_AMT_raw", "COST_AMT_imp", "WEIGHT", "LAST_GRADE", "JUDGE_SEX", "sido"]]
    .head(20)
)

COST_AMT_raw == 0: 0
COST_AMT_imp == 0: 0


,COST_AMT_raw,COST_AMT_imp,WEIGHT,LAST_GRADE,JUDGE_SEX,sido


즉, 원본 금액에도 0원은 없고, 대치 후 금액에도 0원은 없습니다.
따라서 COST_AMT = 0을 결측으로 다시 처리할 필요는 없습니다.

In [51]:
# 0이어도 괜찮은지 확인
df["COST_AMT_raw"] = df["COST_AMT"]

invalid_cost_mask = df["COST_AMT_raw"].isna() | (df["COST_AMT_raw"] <= 0)
df["COST_AMT_missing"] = invalid_cost_mask.astype(int)

In [45]:
# 최종 결측 확인
train_imp["COST_AMT_imp"].isna().sum()

np.int64(0)

In [46]:
final_check_cols = [
    "INSFAT", "TISSUE", "YUKSAK", "FATSAK",
    "WINDEX", "REA", "WGRADE", "BACKFAT", "GROWTH",
    "COST_AMT_imp", "COST_PER_WEIGHT_imp"
]

final_check_cols = [c for c in final_check_cols if c in train_imp.columns]

display(
    train_imp[final_check_cols]
    .isna()
    .sum()
    .sort_values(ascending=False)
)

,0
COST_PER_WEIGHT_imp,1
TISSUE,0
INSFAT,0
YUKSAK,0
FATSAK,0
REA,0
WINDEX,0
WGRADE,0
BACKFAT,0
GROWTH,0


In [47]:
# COST_PER_WEIGHT_imp 1개 남은 결측치 처리 - 어떤 게 남았는지 확인
problem_rows = train_imp[train_imp["COST_PER_WEIGHT_imp"].isna()]

display(
    problem_rows[
        [
            "COST_AMT_raw", "COST_AMT_imp",
            "COST_PER_WEIGHT_raw", "COST_PER_WEIGHT_imp",
            "WEIGHT", "LAST_GRADE", "JUDGE_SEX", "AGE", "sido"
        ]
    ]
)

,COST_AMT_raw,COST_AMT_imp,COST_PER_WEIGHT_raw,COST_PER_WEIGHT_imp,WEIGHT,LAST_GRADE,JUDGE_SEX,AGE,sido
225087,0.0,0.0,NaN,NaN,405.0,1+A,거세,36,경상북도


In [48]:
# ============================================================
# COST_AMT = 0으로 인해 남은 1개 행 처리
# ============================================================

zero_cost_mask = (
    (train_imp["COST_AMT_imp"] == 0) |
    (train_imp["COST_PER_WEIGHT_imp"].isna())
)

display(
    train_imp.loc[
        zero_cost_mask,
        [
            "COST_AMT_raw", "COST_AMT_imp",
            "COST_PER_WEIGHT_raw", "COST_PER_WEIGHT_imp",
            "WEIGHT", "LAST_GRADE", "JUDGE_SEX", "AGE", "sido"
        ]
    ]
)

# LightGBM으로 남은 1개 행 재예측
target_mask = train_imp["COST_PER_WEIGHT_imp"].isna()

X_target = train_imp.loc[target_mask, feature_cols].copy()

pred_cpw = np.expm1(final_cost_model.predict(X_target))
pred_cpw = np.maximum(pred_cpw, 0)

train_imp.loc[target_mask, "COST_PER_WEIGHT_imp"] = pred_cpw

train_imp.loc[target_mask, "COST_AMT_imp"] = (
    train_imp.loc[target_mask, "COST_PER_WEIGHT_imp"]
    * train_imp.loc[target_mask, "WEIGHT"]
)

train_imp.loc[target_mask, "COST_AMT_missing"] = 1
train_imp.loc[target_mask, "COST_AMT_impute_method"] = "lightgbm_cost_per_weight_invalid_zero"

,COST_AMT_raw,COST_AMT_imp,COST_PER_WEIGHT_raw,COST_PER_WEIGHT_imp,WEIGHT,LAST_GRADE,JUDGE_SEX,AGE,sido
225087,0.0,0.0,NaN,NaN,405.0,1+A,거세,36,경상북도


In [49]:
final_check_cols = [
    "INSFAT", "TISSUE", "YUKSAK", "FATSAK",
    "WINDEX", "REA", "WGRADE", "BACKFAT", "GROWTH",
    "COST_AMT_imp", "COST_PER_WEIGHT_imp"
]

display(
    train_imp[final_check_cols]
    .isna()
    .sum()
    .sort_values(ascending=False)
)

print("COST_AMT_imp == 0:", (train_imp["COST_AMT_imp"] == 0).sum())
print("COST_PER_WEIGHT_imp 결측:", train_imp["COST_PER_WEIGHT_imp"].isna().sum())

display(train_imp["COST_AMT_impute_method"].value_counts(dropna=False))
# 끝!

,0
INSFAT,0
TISSUE,0
YUKSAK,0
FATSAK,0
WINDEX,0
REA,0
WGRADE,0
BACKFAT,0
GROWTH,0
COST_AMT_imp,0


COST_AMT_imp == 0: 0
COST_PER_WEIGHT_imp 결측: 0


,count
COST_AMT_impute_method,
observed,1523695
lightgbm_cost_per_weight,885003
lightgbm_cost_per_weight_invalid_zero,1


In [50]:
# 최종 저장
train_imp.to_csv("train_imputed_final.csv", index=False, encoding="utf-8-sig")

In [52]:
from google.colab import files
files.download("train_imputed_final.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [55]:
train_imp.columns

Index(['sido', 'sigungu', 'eupmyeondong', 'stn', 'ABATT_DATE', 'JUDGE_DATE',
       'JUDGE_SEX', 'WEIGHT', 'BACKFAT', 'REA', 'WINDEX', 'WGRADE', 'INSFAT',
       'YUKSAK', 'FATSAK', 'TISSUE', 'GROWTH', 'COST_AMT', 'AGE', 'BIRTH_YMD',
       'CATTLE_NO', 'FARM_UNIQUE_NO', 'LAST_GRADE', 'JUDGE_YEAR', 'age_bin',
       'INSFAT_raw', 'INSFAT_missing', 'INSFAT_impute_method', 'TISSUE_raw',
       'TISSUE_missing', 'TISSUE_impute_method', 'YUKSAK_raw',
       'YUKSAK_missing', 'YUKSAK_impute_method', 'FATSAK_raw',
       'FATSAK_missing', 'FATSAK_impute_method', 'WINDEX_raw',
       'WINDEX_missing', 'WINDEX_impute_method', 'REA_raw', 'REA_missing',
       'REA_impute_method', 'BACKFAT_raw', 'BACKFAT_missing',
       'BACKFAT_impute_method', 'GROWTH_raw', 'GROWTH_missing',
       'GROWTH_impute_method', 'WGRADE_raw', 'WGRADE_missing',
       'WGRADE_impute_method', 'COST_AMT_raw', 'COST_AMT_missing',
       'COST_PER_WEIGHT_raw', 'COST_PER_WEIGHT_imp', 'COST_AMT_imp',
       'COST_AMT_impute